Merge Multiloader into dev

IMS212 opened the pull request:
take two

IMS212 committed:
Start Architectury Port

- Fabric Block Entity Render Data currently removed
- Forge alternative also incomplete

IMS212 committed:
oops

IMS212 committed:
Fix model data issues

IMS212 committed:
Fix buildscript oversights

IMS212 committed:
Fix minor crash/buildscript issues

IMS212 committed:
Replace NeoForge iterateVisibleBlockEntities

IMS212 committed:
Fix issues with the iterator

IMS212 committed:
Fix bugs with empty model data

IMS212 committed:
Fix compatibility with FramedBlocks

IMS212 committed:
Fix the moon clouds

IMS212 committed:
Fix compatibility with Lucent and ModernUI

IMS212 committed:
Fix incorrect fluid color handling

IMS212 committed:
Add support for face hiding

IMS212 committed:
Fix translucency sorting on Fabric API & NeoForge

IMS212 committed:
Fix Bumblezone crash and texture issue

IMS212 committed:
Fix multipart model data not working

IMS212 committed:
Make workarounds not rely on Minecraft

IMS212 committed:
Don't trust the normals

IMS212 committed:
Respect any "override" normals/light values currently in the quad

IMS212 committed:
Remove incorrect info and fill in missing implementations

IMS212 committed:
Make Sodium a SERVICE with JiJ functionality, fix BiggerReactors

IMS212 committed:
Improve normal check and model handling

IMS212 committed:
Include licenses and desktop warn in Forge build

IMS212 committed:
Many fixes to mod metadata

IMS212 committed:
Set a MC version lock

IMS212 committed:
More improvements to multipart models

IMS212 committed:
Initial support for Fabric Rendering API

Co-authored-by: Technici4n <13494793+Technici4n@users.noreply.github.com>
Co-authored-by: PepperCode1 <44146161+PepperCode1@users.noreply.github.com>

IMS212 committed:
Fix more minor issues with port

IMS212 committed:
Mark Indium as provided

IMS212 committed:
Fix multiple bugs with Forge

IMS212 committed:
Multiloader test 1

Stop; it works.

Proper Forge (test 1)

IMS212 committed:
Fix Fabric support

IMS212 committed:
Finally finish the separate buildscript

IMS212 committed:
Remove unneeded implementation

jellysquid3 committed:
Update information about granting licenses

IMS212 committed:
Second implementation of Multiloader buildscript

Fix implementation of runClient

Add archiveClassifier to jar

IMS212 committed:
Fix the Fabric build

IMS212 committed:
Port FRAPI fixes from 1.20.1 branch

IMS212 committed:
Fix model data being checked multiple times

IMS212 committed:
Get proper model data beforehand

IMS212 committed:
Properly include licenses in mod jars

IMS212 committed:
Make Gradle more happy with this configuration

IMS212 committed:
Fix fluid rendering occlusion check

Thanks cortex and moul

IMS212 committed:
Add explicit Immersive Engineering API support

IMS212 committed:
Remove temporary buildscript changes

IMS212 committed:
Fix smooth lighting option not functioning

IMS212 committed:
Support NeoForge AddSectionGeometryEvent

IMS212 committed:
Port to 1.20.5

IMS212 committed:
Clean up platform-specific code

IMS212 committed:
Update to 1.20.6

IMS212 committed:
Merge remote-tracking branch 'refs/remotes/origin/dev' into asfs

# Conflicts:
#	build.gradle.kts
#	common/src/main/java/net/caffeinemc/mods/sodium/mixin/core/model/FaceBakeryMixin.java

IMS212 committed:
Add back removed Mixin during merge

IMS212 committed:
Fix artifacts

IMS212 committed:
Merge branch 'refs/heads/1.20.4/0.6-multiloader' into 1.20.5/0.6-multiloader

# Conflicts:
#	build.gradle.kts
#	neoforge/build.gradle.kts

IMS212 committed:
Update to latest NeoForge for 1.20.6

IMS212 committed:
Provide an auxilary light manager for Forge

IMS212 committed:
Merge remote-tracking branch 'refs/remotes/origin/dev' into 1.20.4/0.6-multiloader

# Conflicts:
#	common/src/main/java/net/caffeinemc/mods/sodium/mixin/features/options/render_layers/LeavesBlockMixin.java

IMS212 committed:
Merge branch 'refs/heads/1.20.4/0.6-multiloader' into 1.20.5/0.6-multiloader

# Conflicts:
#	neoforge/src/main/java/net/caffeinemc/mods/sodium/neoforge/mixin/LevelSliceMixin.java

IMS212 committed:
Fix build and provide temporary fix for Forge item models

MeeniMc committed:
Bugfix: fluid-rendered would occlude water behing translucent blocks (#2468)

This bug was introduced in bc38d5638d391d3dfd1a928f6c7981116a7510f7

IMS212 committed:
Attempt to fix memory leak in chunk meshing

MeeniMc committed:
gh actions: 1.20.6 requires java 21 (#2478)

Signed-off-by: MeeniMc <68366846+MeeniMc@users.noreply.github.com>

MeeniMc committed:
The top face of fluids must not be occluded even if the block above has a full-face bottom. The top fluid face is not coplanar with the bottom full-face, and thus can be seen from sideways, and the top side should be rendered so that underground aquifers and lava pockets render properly in spectator mode. (#2477)

Signed-off-by: MeeniMc <68366846+MeeniMc@users.noreply.github.com>

IMS212 committed:
Fix single-block rendering not using fast path on NeoForge

IMS212 committed:
Refactor normal handling

- Per vertex normals are now always used in place of per-face normals
- Vectors are always normalized according to the PoseStack trustedNormals value

IMS212 committed:
Merge remote-tracking branch 'refs/remotes/origin/dev' into fddf

# Conflicts:
#	build.gradle.kts
#	common/src/api/java/net/caffeinemc/mods/sodium/api/math/MatrixHelper.java
#	common/src/main/java/net/caffeinemc/mods/sodium/client/render/immediate/model/BakedModelEncoder.java
#	common/src/main/java/net/caffeinemc/mods/sodium/client/util/OsUtils.java
#	common/src/main/java/net/caffeinemc/mods/sodium/mixin/core/render/world/LevelRendererMixin.java
#	common/src/main/java/net/caffeinemc/mods/sodium/mixin/features/options/render_layers/LeavesBlockMixin.java
#	common/src/main/java/net/caffeinemc/mods/sodium/mixin/features/render/immediate/matrix_stack/VertexConsumerMixin.java
#	common/src/main/java/net/caffeinemc/mods/sodium/mixin/features/render/world/clouds/LevelRendererMixin.java
#	common/src/main/java/net/caffeinemc/mods/sodium/mixin/features/world/storage/ClientChunkCacheMixin.java
#	common/src/workarounds/java/net/caffeinemc/mods/sodium/client/compatibility/checks/EarlyDriverScanner.java
#	common/src/workarounds/java/net/caffeinemc/mods/sodium/client/compatibility/environment/probe/GraphicsAdapterProbe.java
#	common/src/workarounds/java/net/caffeinemc/mods/sodium/client/compatibility/workarounds/Workarounds.java
#	common/src/workarounds/java/net/caffeinemc/mods/sodium/client/compatibility/workarounds/nvidia/NvidiaWorkarounds.java
#	common/src/workarounds/java/net/caffeinemc/mods/sodium/client/platform/MessageBox.java
#	common/src/workarounds/java/net/caffeinemc/mods/sodium/client/platform/windows/api/msgbox/MsgBoxParamSw.java
#	src/main/java/net/caffeinemc/mods/sodium/client/render/chunk/compile/pipeline/BlockRenderer.java

IMS212 committed:
Merge in changes from dev

IMS212 committed:
Fix incorrect normal getter

IMS212 committed:
Merge all changes

IMS212 committed:
Fix module loading

IMS212 committed:
Merge in translucency fixes

IMS212 committed:
Fix crash when placing string

IMS212 committed:
Update Neo and Gradle scripts

IMS212 committed:
Merge remote-tracking branch 'refs/remotes/origin/dev' into 1.20.5/0.6-multiloader

# Conflicts:
#	common/src/main/java/net/caffeinemc/mods/sodium/mixin/features/render/world/clouds/LevelRendererMixin.java

IMS212 committed:
Apply clouds after NeoForge checks level

IMS212 committed:
Fix impossible check on 1.20.6

This fixes honey rendering on The Bumblezone.

MeeniMc committed:
bugfix: missing accesswidener for fabric (#2466)

IMS212 committed:
Merge remote-tracking branch 'refs/remotes/origin/dev' into 1.20.5/0.6-multiloader

IMS212 committed:
Update NeoForge and allow using DevLogin optionally

IMS212 committed:
1.21 - Part 1

IMS212 committed:
1.21 - Part 3

IMS212 committed:
AO changes and more reliable FAPI dependency

IMS212 committed:
Merge branch 'refs/heads/1.20.5/0.6-multiloader' into 1.21/0.6-multiloader

# Conflicts:
#	neoforge/build.gradle.kts

IMS212 committed:
Update NeoForge

IMS212 committed:
Fix 1.21 loading in production

IMS212 committed:
Fix line rendering

Moulberry committed:
Disable rendering clouds if texture is completely blank (#2536)

IMS212 committed:
Add LWJGL check and (disabled) TFD implementation

IMS212 committed:
Fix chunk vertex consumer

IMS212 committed:
Migrate to ModDevGradle

IMS212 committed:
Use launcher brand to figure out Prism

jellysquid3 committed:
Add early launch drivers for Intel Xe to vendor list

jellysquid3 committed:
Update toolchain versions in README

IMS212 committed:
Do not allow adapter check to crash the game

IMS212 committed:
Fix build

IMS212 committed:
"For Fabric" no more

IMS212 committed:
Fix icon

IMS212 committed:
Add enhanced AO support

douira committed:
[TS] Fix sorting crash when geometry with dot product of -1 is processed

IMS212 committed:
Enhance the enhanced AO support with fixed fluids

IMS212 committed:
Fix glint rendering issue (Indigo PR port)

IMS212 committed:
Merge branch 'refs/heads/1.20.5/0.6-multiloader' into 1.21/0.6-multiloader

# Conflicts:
#	build.gradle.kts
#	neoforge/build.gradle.kts

IMS212 committed:
Fix build

IMS212 committed:
Fix crash when destroying FRAPI models

IMS212 committed:
Fix crash when destroying FRAPI models

IMS212 committed:
Clamp depth values

IMS212 committed:
Clamp depth values

IMS212 committed:
Fix countInjections

IMS212 committed:
Merge branch 'refs/heads/1.20.5/0.6-multiloader' into 1.21/0.6-multiloader

IMS212 committed:
Add support for Parchment mappings

IMS212 committed:
Fix oversight in exterior vertex lighting

IMS212 committed:
Fix multiple issues related to biome colors and pack load order

IMS212 committed:
Remove Immersive Engineering compatibility on 1.20.1

IMS212 committed:
Revert "Use more efficient unpacking in vertex shader"

This reverts commit 17916724

IMS212 committed:
Rewrite platform specific accessors

IMS212 committed:
Port changes to 1.21

IMS212 committed:
Fix Fabric build

IMS212 committed:
Move to FFAPI 1.21

IMS212 committed:
Major fixes to Gradle buildscripts

IMS212 committed:
Run emitBlockQuads for each render type

IMS212 committed:
Clean up inspections and platform code, and delete incorrect code

IMS212 committed:
Changes to model data handling and improved meshing performance on Forge

IMS212 committed:
Fix flipped hidesNeighborState

IMS212 committed:
Fix entity outline colors

IMS212 committed:
Migrate actions to use modern Gradle tooling and update deps

IMS212 committed:
Annotate Github actions

IMS212 committed:
Do not silently crash when incompatible mod is installed

IMS212 committed:
Drastic changes to buildscript

IMS212 committed:
Apply more improvements to NeoForge script

IMS212 committed:
Better mod compatibility for entity rendering

This slightly reduces performance but has more explicit hooks. Mods can now inject into ModelPart.Cube and still render things.

IMS212 committed:
Allow the normal matrix to be changed in between cube renders

IMS212 committed:
Add breaks clause for Embeddium

IMS212 committed:
Use new ModContainer entrypoint (update to NeoForge 21.0.110)

IMS212 committed:
Fix fast clouds crash when going above cloud level

IMS212 committed:
Make sure elements list is 0 immediately following a push

MCRcortex committed:
Include the perspective matrix when marking graph as dirty (#2617)

Co-authored-by: mcrcortex <18544518+MCRcortex@users.noreply.github.com>

IMS212 committed:
Fix incorrect driver parsing code

Fixes #2614

IMS212 committed:
Upgrade NeoForge

IMS212 committed:
Never use the built in color provider if a mod overrides it

Moulberry committed:
Fix delay with FREX Flawless Frames (#2604)

* Fix delay with FREX Flawless Frames due to dependency between building & visibility

* Add iteration limit based on renderDistance

ading2210 committed:
Fix issue #2591 by ignoring null driver names (#2613)

* fix issue #2591 by ignoring null driver names

* add nullable annotation

LostLuma committed:
Fix active option slider labels showing unformatted values (#2615)

* Fix active option slider labels showing unformatted values

* Fix option names not being truncated due to incorrect reported slider width

MCRcortex committed:
Fix for #2586 (#2619)

Co-authored-by: mcrcortex <{ID}+{username}@users.noreply.github.com>

IMS212 committed:
Update plugins and refmap fix

IMS212 committed:
Fix launching with refmaps

IMS212 committed:
Fix incorrect hash code

IMS212 committed:
Merge remote-tracking branch 'origin/1.21/0.6-multiloader' into dev

# Conflicts:
#	build.gradle.kts
#	common/src/main/java/net/caffeinemc/mods/sodium/client/render/immediate/CloudRenderer.java
#	common/src/main/java/net/caffeinemc/mods/sodium/client/render/vertex/VertexFormatDescriptionImpl.java
#	common/src/main/java/net/caffeinemc/mods/sodium/mixin/core/render/immediate/consumer/EntityOutlineGeneratorMixin.java
#	common/src/main/java/net/caffeinemc/mods/sodium/mixin/core/render/world/LevelRendererMixin.java
#	common/src/main/java/net/caffeinemc/mods/sodium/mixin/features/gui/screen/LevelLoadingScreenMixin.java
#	common/src/main/java/net/caffeinemc/mods/sodium/mixin/features/render/immediate/buffer_builder/sorting/MeshDataMixin.java
#	common/src/workarounds/java/net/caffeinemc/mods/sodium/client/platform/MessageBox.java
#	fabric/src/main/java/net/caffeinemc/mods/sodium/fabric/texture/SpriteFinderCache.java
#	src/main/java/net/caffeinemc/mods/sodium/client/render/chunk/compile/buffers/ChunkVertexConsumer.java
#	src/main/java/net/caffeinemc/mods/sodium/mixin/core/render/immediate/consumer/BufferBuilderMixin.java
#	src/main/java/net/caffeinemc/mods/sodium/mixin/features/render/entity/ModelPartMixin.java
#	src/main/java/net/caffeinemc/mods/sodium/mixin/features/render/entity/cull/EntityRendererMixin.java
#	src/main/java/net/caffeinemc/mods/sodium/mixin/features/render/particle/SingleQuadParticleMixin.java
#	src/main/resources/fabric.mod.json

jellysquid3 commented on neoforge/src/main/java/net/caffeinemc/mods/sodium/neoforge/mixin/features/model/WeightedBakedModelMixin.java:
```diff
@@ -0,0 +1,79 @@
+     * @reason Avoid excessive object allocations
+     */
+    @Overwrite
+    public List<BakedQuad> getQuads(@javax.annotation.Nullable BlockState state, @javax.annotation.Nullable Direction face, RandomSource random, ModelData modelData, @org.jetbrains.annotations.Nullable RenderType renderType) {
```
We should look at sending a patch to NeoForge so they aren't using multiple implementations of `@Nullable`.

IMS212 replied:
They currently do not plan to fix this, as Vanilla itself is inconsistent.

jellysquid3 commented on neoforge/src/main/java/net/caffeinemc/mods/sodium/neoforge/mixin/core/model/quad/BakedQuadMixin.java:
```diff
@@ -0,0 +1,134 @@
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+@Mixin(BakedQuad.class)
+public class BakedQuadMixin implements BakedQuadView {
```
How is this different from the Fabric mixin?

IMS212 replied:
It uses a different injection point. This cannot be simply pseudoed in, as the Mixin target method must always exist.

jellysquid3 commented on neoforge/src/main/java/net/caffeinemc/mods/sodium/neoforge/mixin/ResourcePackLoaderMixin.java:
```diff
@@ -0,0 +1,35 @@
+    private static final IModFile SODIUM_FILE = LoadingModList.get().getModFileById("sodium").getFile();
+
+    @Redirect(remap = false, method = "packFinder", at = @At(value = "INVOKE", target = "Ljava/util/Map;entrySet()Ljava/util/Set;"))
+    private static Set<Map.Entry<IModFile, Pack.ResourcesSupplier>> changeSet(Map<IModFile, Pack.ResourcesSupplier> instance) {
```
This looks to be a patch for mods that also overwrite the leaves texture. Can you document the reasoning here for future reference?

IMS212 replied:
Added. (It's above the preview)

jellysquid3 commented on common/src/main/java/net/caffeinemc/mods/sodium/client/services/SodiumModelData.java:
```diff
@@ -0,0 +1,8 @@
+/**
+ * Template class for the platform's model data.
+ */
+public interface SodiumModelData {
```
This should be documented to explain that it's used to shadow the Forge types in a Fabric environment.

IMS212 replied:
Added

jellysquid3 commented on neoforge/src/main/resources/META-INF/neoforge.mods.toml:
```diff
@@ -0,0 +1,43 @@
+modLoader = "javafml"
+loaderVersion = "*"
+license = "Polyform-Shield-1.0"
```
Should be `Polyform-Shield-1.0.0`.

jellysquid3 commented on neoforge/src/main/resources/META-INF/neoforge.mods.toml:
```diff
@@ -0,0 +1,43 @@
+
+logoFile = "sodium-icon.png" #optional
+
+authors = "JellySquid"
```
We should also include a secondary authors field (maybe credits?) so that our contributors are listed.

jellysquid3 commented on neoforge/src/main/java/net/caffeinemc/mods/sodium/neoforge/mixin/ClientHooksMixin.java:
```diff
@@ -0,0 +1,17 @@
+    // TODO: This is not a proper fix at all. Figure out why Forge is causing the UV shrink ratio to go insane.
+    @Redirect(method = "fixItemModelSeams", at = @At(value = "INVOKE", target =
+            "Lnet/minecraft/client/renderer/texture/TextureAtlasSprite;uvShrinkRatio()F"))
+    private static float alterUvShrinkRatio(TextureAtlasSprite sprite) {
```
Does this still need to be investigated before release?

jellysquid3 commented on neoforge/src/main/java/net/caffeinemc/mods/sodium/neoforge/level/NeoForgeLevelAccess.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    @Override
+    public @Nullable Object getLightManager(LevelChunk chunk, SectionPos pos) {
```
Can the type erasure be avoided here?

IMS212 replied:
Replaced with `SodiumAuxilaryLightManager`, similar to `SodiumModelData`.

jellysquid3 commented on neoforge/src/main/java/net/caffeinemc/mods/sodium/neoforge/level/NeoForgeLevelAccess.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.function.Function;
+
+public class NeoForgeLevelAccess implements PlatformLevelAccess {
+    private static final ThreadLocal<PoseStack> emptyStack = ThreadLocal.withInitial(PoseStack::new);
```
Why does this need to be cached in a thread local? The allocation rate is only once per chunk mesh event. Keeping it around increases the risk that a mod could break the pose stack persistently between chunks and would make debugging harder. And if the stack gets very large, that memory will be kept around.

IMS212 replied:
Removed in favor of a `new PoseStack()`.

jellysquid3 commented on neoforge/src/main/java/net/caffeinemc/mods/sodium/neoforge/block/NeoForgeBlockAccess.java:
```diff
@@ -0,0 +1,62 @@
+import org.jetbrains.annotations.Nullable;
+
+public class NeoForgeBlockAccess implements PlatformBlockAccess {
+    private static final TriState[] TRI_STATES = new TriState[] {
```
Is `TriState.of` not sufficient? If there is some platform difference here, let's map each platform's tri-state to a special `ShadeMode` or such.

IMS212 replied:
Replaced with `AmbientOcclusionMode`.

jellysquid3 commented on neoforge/src/main/java/net/caffeinemc/mods/sodium/neoforge/NeoForgeInfoAccess.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    @Override
+    public boolean isFlawlessFramesActive() {
```
Why is this in platform-specific code? The Fabric mod initializer should scan the mods and set up the global Flawless Frames consumers.

jellysquid3 commented on common/src/main/java/net/caffeinemc/mods/sodium/client/services/PlatformInfoAccess.java:
```diff
@@ -0,0 +1,46 @@
+
+import java.nio.file.Path;
+
+public interface PlatformInfoAccess {
```
"PlatformInfo" is a big vague here, it's more about the runtime environment info.

jellysquid3 commented on common/src/main/java/net/caffeinemc/mods/sodium/client/services/PlatformLevelAccess.java:
```diff
@@ -0,0 +1,79 @@
+import java.util.List;
+import java.util.function.Function;
+
+public interface PlatformLevelAccess {
```
This should be split into render-specific functionality (`PlatformLevelRenderHooks`, maybe) and world-query functionality (what the file is currently labeled.)

jellysquid3 commented on common/src/main/java/net/caffeinemc/mods/sodium/client/services/PlatformLevelAccess.java:
```diff
@@ -0,0 +1,79 @@
+     * @param origin The origin of the current chunk
+     * @return Any NeoForge chunk renderers to run
+     */
+    List<?> getExtraRenderers(Level level, BlockPos origin);
```
Renderers for what?

jellysquid3 commented on common/src/main/java/net/caffeinemc/mods/sodium/client/services/PlatformLevelAccess.java:
```diff
@@ -0,0 +1,79 @@
+     * @param lightPipelineProvider The current {@code LightPipelineProvider}.
+     * @return A new fluid renderer.
+     */
+    FluidRenderer createPlatformFluidRenderer(ColorProviderRegistry colorRegistry, LightPipelineProvider lightPipelineProvider);
```
Maybe pull this out into a separate `FluidRendererProvider` interface.

IMS212 replied:
Would implementing this in the FluidRenderer itself be smart?

jellysquid3 commented on common/src/main/java/net/caffeinemc/mods/sodium/client/services/PlatformLevelAccess.java:
```diff
@@ -0,0 +1,79 @@
+     * Should be run in the top of the Vanilla fluid renderer. If true, cancel the Vanilla render.
+     * @return if the vanilla rendering should be cancelled.
+     */
+    boolean tryRenderFluid();
```
Why does this exist? Is there no way to get it out of the platform interface? It's incredibly fragile.

IMS212 replied:
Replaced with FluidRendering.render, a Fabric API designed for this.

jellysquid3 commented on common/src/main/java/net/caffeinemc/mods/sodium/client/services/PlatformLevelAccess.java:
```diff
@@ -0,0 +1,79 @@
+     * @param typeToConsumer A consumer that converts render types to vertex consumers
+     * @param slice The current level slice
+     */
+    void renderAdditionalRenderers(List<?> renderers, Function<RenderType, VertexConsumer> typeToConsumer, LevelSlice slice);
```
This really needs a better name. Maybe `collectAdditionalChunkGeometry` or something.

jellysquid3 commented on common/src/main/java/net/caffeinemc/mods/sodium/client/services/PlatformTextureAccess.java:
```diff
@@ -0,0 +1,19 @@
+     * @param texV The V coordinate of the texture.
+     * @return The sprite in the location in the block atlas.
+     */
+    TextureAtlasSprite findInBlockAtlas(float texU, float texV);
```
I think we should instead provide a `SpriteFinder getSpriteFinderForBlockAtlas()` or something, and have each platform provide an implementation which supports queries. Why is the sprite finder platform-specific, anyways?

jellysquid3 commented on common/src/main/java/net/caffeinemc/mods/sodium/client/services/PlatformBlockAccess.java:
```diff
@@ -0,0 +1,91 @@
+     * @return The specialized render data for this block entity. If the platform does not support it or there is no data, null.
+     */
+    @Nullable
+    Object getBlockEntityData(BlockEntity blockEntity);
```
This should probably be moved into the platform code for accessing level data.

jellysquid3 commented on common/src/main/java/net/caffeinemc/mods/sodium/client/services/PlatformBlockAccess.java:
```diff
@@ -0,0 +1,91 @@
+     * @param shade If directional lighting should be added.
+     * @return the block shade
+     */
+    float getAccurateBlockShade(ModelQuadView quad, BlockAndTintGetter level, boolean shade);
```
Why is it accurate? Is this render code or level code?

IMS212 replied:
This uses the quad’s normal vector to calculate the shade value. What would an appropriate name be?

IMS212 replied:
Renamed to `getNormalVectorShade`.

jellysquid3 commented on common/src/main/java/net/caffeinemc/mods/sodium/client/render/immediate/CloudRenderer.java:
```diff
@@ -337,23 +335,15 @@ private static void emitCellGeometry2D(VertexBufferWriter writer, int faces, int
-            // -Y
-            if (CloudFaceSet.contains(faces, CloudFace.NEG_Y)) {
-                int mixedColor = ColorMixer.mul(color, CloudFace.POS_Y.getColor());
+            int mixedColor = ColorMixer.mul(color, CloudFace.POS_Y.getColor());
```
Why are these faces always rendered now? If the goal is to avoid an empty buffer, this is not a good way to do it, since it completely eliminates our optimizations.

Worst case, we can update the method signature to return a boolean indicating whether geometry was added to the buffer.

IMS212 replied:
Removed for now.

IMS212 committed:
Revert "Fix fast clouds crash when going above cloud level"

This reverts commit 6fad4e3be04042d44a04e74085c3a498c1109b1a.

jellysquid3 commented on common/src/main/java/net/caffeinemc/mods/sodium/client/render/frapi/material/MaterialViewImpl.java:
```diff
@@ -0,0 +1,107 @@
+
+    @Override
+    public TriState ambientOcclusion() {
+        return TRI_STATES[(bits & AO_MASK) >>> AO_BIT_OFFSET];
```
Why not use `TriState.of(bool)` here?

IMS212 replied:
I’m pretty sure this can return DEFAULT.

IMS212 committed:
Fix multiple review issues

- Move Flawless Frames integration
- Rename and modify many services
- Made SpriteFinder independent of Fabric API
- Added more explicit interfaces

IMS212 committed:
Fix more review issues

IMS212 committed:
Refactor the platform fluid renderers

IMS212 committed:
Merge remote-tracking branch 'origin/dev' into take2

# Conflicts:
#	common/src/workarounds/java/net/caffeinemc/mods/sodium/client/compatibility/checks/ModuleScanner.java

jellysquid3 commented on common/src/api/java/net/caffeinemc/mods/sodium/api/util/ColorMixer.java:
```diff
@@ -59,4 +59,19 @@ public static int mul(int a, int b) {
+    /**
+     * Multiplies the 32-bit colors with one component
+     */
+    public static int mulSingle(int a, byte b) {
```
This seems like it would be broken. Java uses signed bytes. So the valid range is between `-127` and `128`, and using it in multiplication results in it being sign-extended...

IMS212 replied:
Fixed by replacing it with int. This isn't actively used anywhere (yet).

jellysquid3 commented on common/src/api/java/net/caffeinemc/mods/sodium/api/util/NormI8.java:
```diff
@@ -114,4 +115,8 @@ public static boolean isOpposite(int normA, int normB) {
         return normAX == -normBX && normAY == -normBY && normAZ == -normBZ;
     }
+
+    public static Vector3f unpack(int i, Vector3f output) {
```
i -> packed

jellysquid3 commented on common/src/api/java/net/caffeinemc/mods/sodium/api/math/MatrixHelper.java:
```diff
@@ -129,6 +138,15 @@ public static void rotateZYX(PoseStack.Pose matrices, float angleZ, float angleY
                 .rotateZYX(angleZ, angleY, angleX);
     }
 
+    /**
```
This seems like a dangling Javadoc comment? Where is the associated function?

jellysquid3 commented on common/src/api/java/net/caffeinemc/mods/sodium/api/math/MatrixHelper.java:
```diff
@@ -17,17 +17,26 @@
      * @return The transformed normal vector (in packed format)
      */
-    public static int transformNormal(Matrix3f mat, float x, float y, float z) {
+    public static int transformNormal(Matrix3f mat, boolean skipNormalization, float x, float y, float z) {
```
Probably, this function should be split into two (one with and without the `skipNormalization` parameter) so the branch only gets evaluated when it's actually necessary.

jellysquid3 commented on common/src/main/java/net/caffeinemc/mods/sodium/client/data/config/MixinConfig.java:
```diff
@@ -129,47 +128,23 @@ private void readProperties(Properties props) {
-                    LOGGER.warn("Mod '{}' contains invalid Sodium option overrides, ignoring", meta.getId());
-                    continue;
-                }
+    public abstract void applyModOverrides();
```
I don't think we want to abstract the MixinPlugin itself, but rather there should be a service for obtaining the overrides from mod manifests.

jellysquid3 commented on common/src/main/java/net/caffeinemc/mods/sodium/client/model/color/DefaultColorProviders.java:
```diff
@@ -30,7 +26,7 @@ private GrassColorProvider() {
         @Override
         protected int getColor(LevelSlice slice, int x, int y, int z) {
-            return slice.getColor(BiomeColorSource.GRASS, x, y, z);
+            return BiomeColors.getAverageGrassColor(slice, blockPosHolder.get().set(x, y, z));
```
Why did this change?

IMS212 replied:
This changed to fix mod compatibility with a few mods, including Hydrological. It does not seem to incur any heavy cost, minus the usage of the BlockPos in the below comment.

jellysquid3 commented on common/src/main/java/net/caffeinemc/mods/sodium/client/model/color/DefaultColorProviders.java:
```diff
@@ -17,9 +15,7 @@ public static ColorProvider<BlockState> adapt(BlockColor color) {
-    public static ColorProvider<FluidState> adapt(FluidRenderHandler handler) {
-        return new FabricFluidAdapter(handler);
-    }
+    private static final ThreadLocal<BlockPos.MutableBlockPos> blockPosHolder = ThreadLocal.withInitial(BlockPos.MutableBlockPos::new);
```
The thread local lookup here is going to be really slow when fetching colors during chunk meshing. Why did it need to be added?

IMS212 replied:
See the comment above; should it be replaced with a new BlockPos() for each quad?

IMS212 replied:
Actually, I have a new idea to replace this, but it may be a bit messy; providing a scratch BlockPos from the BlockRenderer.

jellysquid3 commented on common/src/main/java/net/caffeinemc/mods/sodium/client/model/light/LightPipeline.java:
```diff
@@ -18,6 +18,7 @@ public interface LightPipeline {
+     * @param isFluid True if the block is being drawn by the fluid renderer
      */
-    void calculate(ModelQuadView quad, BlockPos pos, QuadLightData out, Direction cullFace, Direction lightFace, boolean shade);
+    void calculate(ModelQuadView quad, BlockPos pos, QuadLightData out, Direction cullFace, Direction lightFace, boolean shade, boolean isFluid);
```
Might make more sense to pass the `BlockState` or `FluidState` here instead of a boolean called `isFluid`.

IMS212 committed:
Apply requested changes
